auth_totp/models/res_users.py

```diff
@@ -50,53 +56,38 @@ def _check_enabled_with_authenticator(self):
 
     @api.model
     def check_credentials(self, password):
-        try:
+        """Add MFA logic to core authentication process
+
+        Overview:
+            * If user does not have MFA enabled, defer to parent logic
+            * If user has MFA enabled and has gone through MFA login process
+              this session or has correct device cookie, defer to parent logic
+            * If neither of these is true, call parent logic. If successful,
+              prevent auth while updating session to indicate that MFA login
+              process can now commence
+        """
+        user_model_sudo = self.sudo()
+        user = user_model_sudo.search([('id', '=', self.env.uid)])
+
+        if not user.mfa_enabled:
             return super(ResUsers, self).check_credentials(password)
-        except AccessDenied:
-            user = self.sudo().search([
-                ('id', '=', self.env.uid),
-                ('mfa_login_token', '=', password),
-            ])
-            user._user_from_mfa_login_token_validate()
-
-    @api.multi
-    def generate_mfa_login_token(self, lifetime_mins=15):
-        char_set = string.ascii_letters + string.digits
-
-        for record in self:
-            record.mfa_login_token = ''.join(
-                random.SystemRandom().choice(char_set) for __ in range(20)
-            )
-
-            expiration = datetime.now() + timedelta(minutes=lifetime_mins)
-            record.mfa_login_token_exp = fields.Datetime.to_string(expiration)
-
-    @api.model
-    def user_from_mfa_login_token(self, token):
-        if not token:
-            raise MfaTokenInvalidError(_(
-                'Your MFA login token is not valid. Please try again.'
-            ))
-
-        user = self.search([('mfa_login_token', '=', token)])
-        user._user_from_mfa_login_token_validate()
-
-        return user
-
-    @api.multi
-    def _user_from_mfa_login_token_validate(self):
-        try:
-            self.ensure_one()
-        except ValueError:
-            raise MfaTokenInvalidError(_(
-                'Your MFA login token is not valid. Please try again.'
-            ))
 
-        token_exp = fields.Datetime.from_string(self.mfa_login_token_exp)
-        if token_exp < datetime.now():
-            raise MfaTokenExpiredError(_(
-                'Your MFA login token has expired. Please try again.'
-            ))
+        if request:
+            if request.session.get('mfa_login_active') == user.id:
```

Comment:
`mfa_login_active` string key seems pretty integral to things, and is repeated in a lot of spots. Typically I try to make strings into constants once I've used them 3-4 times (not counting tests). Do we have a good place to do this? Maybe in MFA model?

Reply:
This key is only used in two places outside of tests, and any constant used to replace it will need to be just as verbose, so I don't think anything is gained by doing that.